### PR TITLE
backport:  merge bitcoin#15748, #18541, #19848, #17407 [redo], #14384, #17477, #19826, #19556, #19854  -- deglobalisation of mempool

### PR DESCRIPTION
**Pull request by knst**
This pull request is part of deglobalisation of mempool.
Merged backports:
 - bitcoin#15748
 - bitcoin#18541
 - bitcoin#19848
 - [redo] fix: re-run a script from backport bitcoin#17407
 - bitcoin#14384
 - bitcoin#17477
 - bitcoin#19826
 - bitcoin#19556
 - bitcoin#19854

This backports does not work correctly without #5174 and #5175

## How Has This Been Tested?
Run unit/functional test and introduce no breaking changes.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
IMO, the first two non-backport commits should be in another PR

**Comment by knst** [comment]
> IMO, the first two non-backport commits should be in another PR

Please, review #5175 and #5174

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — APPROVED** [review]
utACK

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/85741ebd7862168987128db1aba9980c5084990e..81931f9380846928a4b927ae165bc9a101405307) the bc-bp-mempool branch from[ 85741eb ](https://github.com/dashpay/dash/commit/85741ebd7862168987128db1aba9980c5084990e)to 81931f9

Resolved conflicts with net_processing's refactorins in develop branch

File: src/coinjoin/client.cpp [hunk 1]
```diff
@@ -134,7 +134,7 @@ void CCoinJoinClientQueueManager::ProcessDSQueue(const CNode& peer, CDataStream&
     }
 }
 
-void CCoinJoinClientManager::ProcessMessage(CNode& peer, CConnman& connman, std::string_view msg_type, CDataStream& vRecv)
+void CCoinJoinClientManager::ProcessMessage(const CTxMemPool& mempool, CNode& peer, CConnman& connman, std::string_view msg_type, CDataStream& vRecv)
```
Comment (kittywhiskers) [inline src/coinjoin/client.cpp:137]: nit, in line with the argument order of all other `ProcessMessage`s, consider changing this to `CNode& peer, CConnman& connman, const CTxMemPool& mempool, std::string_view msg_type, CDataStream& vRecv`

File: src/coinjoin/client.cpp [hunk 2]
```diff
@@ -153,12 +153,12 @@ void CCoinJoinClientManager::ProcessMessage(CNode& peer, CConnman& connman, std:
...
         LOCK(cs_deqsessions);
         for (auto& session : deqSessions) {
-            session.ProcessMessage(peer, connman, msg_type, vRecv);
+            session.ProcessMessage(mempool, peer, connman, msg_type, vRecv);
         }
     }
 }
 
-void CCoinJoinClientSession::ProcessMessage(CNode& peer, CConnman& connman, std::string_view msg_type, CDataStream& vRecv)
+void CCoinJoinClientSession::ProcessMessage(const CTxMemPool& mempool, CNode& peer, CConnman& connman, std::string_view msg_type, CDataStream& vRecv)
```
Comment (kittywhiskers) [inline src/coinjoin/client.cpp:161]: nit, see above

File: src/coinjoin/client.h [hunk 3]
```diff
@@ -187,6 +188,7 @@ class CCoinJoinClientManager
     bilingual_str strAutoDenomResult;
 
     CWallet& mixingWallet;
+//    CTxMemPool& mempool;
```
Comment (kittywhiskers) [inline src/coinjoin/client.h:191]: nit, remove unused declaration

**Review by UdjinM6 — DISMISSED** [review]
re-utACK

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/280a8a34eba49036136b89308adfcd1dd47ac8ec..a0ce90536ac564dbc2ad92802b141a4a1887c6ca) the bc-bp-mempool branch from[ 280a8a3 ](https://github.com/dashpay/dash/commit/280a8a34eba49036136b89308adfcd1dd47ac8ec)to a0ce905
Resolved conflict with #5178

File: src/wallet/salvage.cpp [hunk 4]
```diff
@@ -6,6 +6,7 @@
 #include <fs.h>
 #include <streams.h>
 #include <util/translation.h>
+#include <validation.h>
```
Comment (UdjinM6) [inline src/wallet/salvage.cpp:9]: 19556: I think this is no longer needed after 5178 merge
```suggestion
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by UdjinM6** [comment]
pls squash [d77e96d](https://github.com/dashpay/dash/pull/5169/commits/d77e96d5eeadca1e0f34cffada5111b01784f46d) into [6e878be](https://github.com/dashpay/dash/pull/5169/commits/6e878becf2ed37ce980ccf8518bda9bb8d807986) (19556)

**Review by UdjinM6 — DISMISSED** [review]
re-utACK

**Review by UdjinM6 — APPROVED** [review]
re-utACK

**Review by kittywhiskers — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**